### PR TITLE
Show total shared annotations in the group and user side bar.

### PR DESCRIPTION
**Pull request by sheetaluk**
Fixes: https://github.com/hypothesis/h/issues/4168

### CODE REVIEW

**Comment by nickstenning** [comment]
I see @chdorner has already been tagged as a reviewer so I'll let him do the honours, but I'll just note that we should probably wrap the count functions up in a service, and not as properties on the User and Group models -- as per [ADR 2](https://github.com/hypothesis/h/blob/master/docs/arch/adr-002.rst) (use services for business logic which touches multiple models, in order to avoid circular dependencies, as well as general model bloat).

File: tests/h/models/user_test.py [hunk 1]
```diff
@@ -289,6 +289,19 @@ def test_it_filters_by_authority(self, db_session, users):
...
 
+    def test_annotation_count_returns_count_of_shared_annotations_for_user(self, db_session, users):
+        user = users['meredith']
+
+        for i in range(50):
+            annotation(db_session,
+                       document(db_session,
+                                'http://example.com/document_' + str(i)),
+                       user.userid,
+                       shared=True)
```
Comment (chdorner) [inline tests/h/models/user_test.py]: I think you can just use the factories here, which saves you quite some typing and will be more future proof when we change the annotation/document models.
Which means that you won't need the annotation and document fixture but can just do `factories.Annotation(user=user.userid, shared=True)` in the test above where you use it.

File: h/accounts/services.py [hunk 2]
```diff
@@ -115,6 +115,16 @@ def login(self, username_or_email, password):
...
         return None
 
+    def annotation_count(self, userid):
+        """
+        Return the count of shared annotations for this user.
+        """
+
+        return (
+            self.session.query(Annotation)
+            .filter_by(userid=userid, shared=True)
```
Comment (chdorner) [inline h/accounts/services.py]: Ah! One thing I totally missed in the last review (sorry), but just noticed now when I tested is that we have to filter by `groupid='__world__'` as well in this case. As we don't want to expose how many annotations a user made in other groups. This also means that the test should create a bunch of annotations in a group that is not `__world__` to test that we don't count them.

**Comment by sheetaluk** [comment]
@chdorner ok, all done. ready for re-review :)

**Comment by ajpeddakotla** [comment]
@chdorner thanks for catching the cases I failed to mention - not counting private annotations or annotations made in other groups.

**Comment by ajpeddakotla** [comment]
So after talking @sheetaluk, there needs to be a finer grained approach to counts of annotations that we are showing in both the sidebar and the search results. 

So if we breakdown the different types of annotations that can be viewed, and who they can be viewed by, it looks something like this (for now - this will change with publisher groups). 

Empty |Private | Shared | Public
------ | ------ | ------ | ------
Me | Y | Y | Y
Person I am in one or more groups with | N | Y (but only in groups that I am a part of) | Y
Stranger | N | N | Y

Right now, what's been implemented - we're only showing public annotations which are not private or shared. 

What we'd like to do, because showing the finer detail helps better inform the user of the types of annotations they are making is to: 

**1. In the sidebar:** 
Breakdown the count to include (for users viewing their own profile):
Shared annotations: 
Private annotations: 
Public annotations: 

For users viewing someone else's profile: 
Public annotations: 
Shared annotations (if they share a common group with that user)

**2. In the search results:** 
If I'm viewing my own profile. 
Shared
Private
Public annotations (if I'm viewing my own profile)

If I'm viewing someone else's annotations
Shared (if they share a common group with me)
Public

**Pull request merged**